File: leetcode/top150/26-remove-duplicates-sorted-array/c++/solution.cpp

#include "templates/solution.hpp"

using namespace solution;

template<>
void Solution::solution(std::vector<std::shared_ptr<output<int>>>& outputs,
                        std::vector<std::shared_ptr<input<std::vector<int>>>>& inputs) {
    
    // Declare variables
    auto [nums] = inputs[0]->inputs;
    int c_val = nums[0];
    int c = 1;
    size_t i = 1;
    std::tuple<int> count;

    // Loop through array and remove duplicates while incrementing
    while (i < nums.size()) {
        // Remove occurence of duplicate
        if (c_val == nums[i]) {
            auto it = nums.begin() + i;
            nums.erase(it);
            continue;
        }
        // Update variables since c_val < nums[i]
        c_val = nums[i];
        ++c;
        ++i;

    };

    // Update input
    // This is of constant complexity N/A
    std::get<0>(inputs[0]->inputs).swap(nums);

    count = std::make_tuple(c);
    outputs[0]->outputs.swap(count);
}

File: leetcode/top150/26-remove-duplicates-sorted-array/python/solution.py
from templates.python.solution import SolutionBase

class Solution(SolutionBase):

    def __init__(self, *inputs):
        super().__init__(*inputs)

    def solution(self):
        c_val = self.inputs[0][0]
        c = 1
        i = 1

        while (i < len(self.inputs[0])):
            if (c_val == self.inputs[0][i]):
                self.inputs[0].pop(i)
                continue
            c_val = self.inputs[0][i]
            c = c + 1
            i = i + 1
        
        self.logger.info(f'Output is {c}')
        return c
